### PR TITLE
 spirv-fuzz: Transformation and fuzzer pass for changing memory scope

### PR DESCRIPTION
**Pull request by Mostafa-ashraf19**
This PR Fixes Issue #4451.

### CODE REVIEW

File: source/fuzz/fuzzer_pass_changing_memory_scope.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      [this](opt::Function* /* unused */, opt::BasicBlock* /* unused */,
+             opt::BasicBlock::iterator inst_it,
+             const protobufs::InstructionDescriptor& instruction_descriptor) {
+        if (!GetFuzzerContext()->ChoosePercentage(
+                GetFuzzerContext()->GetChanceOfChangingMemoryScope())) {
+          return;
+        }
+
+        // Instruction must be atomic instruction only.
+        if (!TransformationChangingMemoryScope::IsAtomicInstruction(
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_changing_memory_scope.cpp:45]: As discussed in our meeting, we should include barrier instructions. Perhaps rename `IsAtomicInstruction` to `HasMemorySemanticsOperand`.

File: source/fuzz/fuzzer_pass_changing_memory_scope.cpp [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+            {static_cast<uint32_t>(memory_scope_new_value)},
+            FindOrCreateIntegerType(32, GetFuzzerContext()->ChooseEven()),
+            false);
+
+        ApplyTransformation(TransformationChangingMemoryScope(
+            instruction_descriptor, new_memory_scope_id));
+      });
+}
+
+std::vector<SpvScope> FuzzerPassChangingMemoryScope::GetSuitableNewMemoryScope(
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_changing_memory_scope.cpp:81]: GetSuitableNewMemoryScope -> GetSuitableNewMemoryScopes

File: source/fuzz/fuzzer_pass_changing_memory_scope.cpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      SpvScopeCrossDevice, SpvScopeDevice, SpvScopeWorkgroup, SpvScopeSubgroup,
+      SpvScopeInvocation};
+
+  // std::remove_if does not actually remove any elements from the vector; it
+  // just reorders them to the end of the vector.
+  auto reordered_memory_semantics = std::remove_if(
+      potential_new_scopes_orders.begin(), potential_new_scopes_orders.end(),
+      [old_memory_scope_value](SpvScope memory_scope) {
+        return (old_memory_scope_value == memory_scope ||
+                memory_scope > old_memory_scope_value);
```
Comment (paulthomson) [inline source/fuzz/fuzzer_pass_changing_memory_scope.cpp:93]: Use TransformationChangingMemoryScope::IsValidScope.

File: source/fuzz/protobufs/spvtoolsfuzz.proto [hunk 4]
```diff
@@ -1296,6 +1297,19 @@ message TransformationAdjustBranchWeights {
 
 }
 
+message TransformationChangingMemoryScope {
+
+  // This transformation is responsible for strengthens the memory scope
+  // operand within an atomic instruction.
```
Comment (paulthomson) [inline source/fuzz/protobufs/spvtoolsfuzz.proto:1303]: Change to: Widens a memory scope operand within an instruction.

File: source/fuzz/transformation_changing_memory_scope.cpp [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+  auto needed_instruction =
+      FindInstruction(message_.needed_instruction(), ir_context);
+  if (!needed_instruction) {
+    return false;
+  }
+
+  // Instruction must be atomic instruction only.
+  if (!IsAtomicInstruction(needed_instruction->opcode())) {
+    return false;
+  }
```
Comment (paulthomson) [inline source/fuzz/transformation_changing_memory_scope.cpp:46]: See your other PR and follow the same approach. We do not need to check the opcode in IsApplicable.

File: source/fuzz/transformation_changing_memory_scope.cpp [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+bool TransformationChangingMemoryScope::IsValidScope(
+    SpvScope new_memory_scope_value, SpvScope old_memory_scope_value) {
+  switch (new_memory_scope_value) {
+    case SpvScopeCrossDevice:
+    case SpvScopeDevice:
+    case SpvScopeWorkgroup:
+    case SpvScopeSubgroup:
+    case SpvScopeInvocation:
+      if (new_memory_scope_value > old_memory_scope_value) {
```
Comment (paulthomson) [inline source/fuzz/transformation_changing_memory_scope.cpp:171]: new_memory_scope_value >= old_memory_scope_value

File: source/fuzz/transformation_changing_memory_scope.cpp [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto needed_instruction =
+      FindInstruction(message_.needed_instruction(), ir_context);
+
+  uint32_t needed_index =
+      GetMemoryScopeInOperandIndex(needed_instruction->opcode());
+
+  needed_instruction->SetInOperand(needed_index,
+                                   {message_.memory_scope_new_value_id()});
+}
```
Comment (paulthomson) [inline source/fuzz/transformation_changing_memory_scope.cpp:102]: I think you need to update the def use manager and call `set_instr_block`, or just clear all analyzes.

File: source/fuzz/transformation_changing_memory_scope.cpp [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+      return true;
+
+    default:
+      return false;
+  }
+}
+
+bool TransformationChangingMemoryScope::IsValidScope(
+    SpvScope new_memory_scope_value, SpvScope old_memory_scope_value) {
+  switch (new_memory_scope_value) {
```
Comment (paulthomson) [inline source/fuzz/transformation_changing_memory_scope.cpp:165]: If old_memory_scope_value is not one of the values below, we should return false.

**Review by paulthomson — COMMENTED** [review]
Looks good. I left a few comments.

File: source/fuzz/transformation_changing_memory_scope.cpp (hunk 6 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_changing_memory_scope.cpp:171]: Done.

File: source/fuzz/protobufs/spvtoolsfuzz.proto (hunk 4 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/protobufs/spvtoolsfuzz.proto:1303]: Done.

File: source/fuzz/fuzzer_pass_changing_memory_scope.cpp (hunk 3 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/fuzzer_pass_changing_memory_scope.cpp:93]: Done.

File: source/fuzz/fuzzer_pass_changing_memory_scope.cpp (hunk 2 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/fuzzer_pass_changing_memory_scope.cpp:81]: Done.

File: source/fuzz/transformation_changing_memory_scope.cpp (hunk 8 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_changing_memory_scope.cpp:165]: Done.

File: source/fuzz/fuzzer_pass_changing_memory_scope.cpp (hunk 1 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/fuzzer_pass_changing_memory_scope.cpp:45]: Done. I believe here you meant `HasMemoryScopesOperand`.

Comment (paulthomson) [reply source/fuzz/fuzzer_pass_changing_memory_scope.cpp:45]: Yes, oops. I think `HasMemoryScopeOperand` would be good (remove "s").

Comment (Mostafa-ashraf19) [reply source/fuzz/fuzzer_pass_changing_memory_scope.cpp:45]: Done.

File: source/fuzz/transformation_changing_memory_scope.cpp (hunk 5 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_changing_memory_scope.cpp:46]: Done.

File: source/fuzz/transformation_changing_memory_scope.cpp (hunk 7 above)
Comment (Mostafa-ashraf19) [reply source/fuzz/transformation_changing_memory_scope.cpp:102]: Done.

**Pull request closed without merging**